### PR TITLE
✨ Integrate kube-state-metrics and CR config into tilt.

### PR DESCRIPTION
**Pull request by chrischdi**
**What this PR does / why we need it**:

* Adds `kube-state-metrics` to `hack/observability` which uses the `kube-state-metrics` helm chart and adds configuration for CAPI CRs
* Integrates into tilt

Currently not added metrics:
- `*_labels`

TODOs:

- [x] Use new kube-state-metrics release as soon as it got published
- [x] ~~Split CR configuration to multiple files if https://github.com/kubernetes/kube-state-metrics/pull/1810 gets merged~~
    - won't implement for now. Split could be done in future when KSM provides a CR for configuration purposes.
- [x] Verify if all expected metrics are configured (except `_labels`)
- [x] update proposal for `*_spec_paused` and `*_annotation_paused` metrics
- [x] Provide docs
- [ ] review kcp and md `*_status_replicas_*` metrics and unify

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Part of #6458

### CODE REVIEW

File: Tiltfile [hunk 1]
```diff
@@ -354,6 +354,10 @@ def deploy_observability():
         k8s_yaml(read_file("./.tiltbuild/yaml/prometheus.observability.yaml"), allow_duplicates = True)
         k8s_resource(workload = "prometheus-server", new_name = "prometheus", port_forwards = "9090", extra_pod_selectors = [{"app": "prometheus"}], labels = ["observability"])
 
+    if "capi-state-metrics" in settings.get("deploy_observability", []):
+        k8s_yaml(read_file("./.tiltbuild/yaml/capi-state-metrics.observability.yaml"), allow_duplicates = True)
+        k8s_resource(workload = "capi-state-metrics", new_name = "capi-state-metrics", port_forwards = "8081:8080", extra_pod_selectors = [{"app": "kube-state-metrics"}], labels = ["observability"])
```
Comment (chrischdi) [inline Tiltfile:359]:
```suggestion
        k8s_resource(workload = "capi-state-metrics", new_name = "capi-state-metrics", port_forwards = "8081:8080", extra_pod_selectors = [{"app": "kube-state-metrics"}], labels = ["observability"])
```

TODO: decide on which port to use, I think we should not block 8080?!

File: hack/observability/capi-state-metrics/kube-state-metrics/values.yaml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+collectors:
+- clusters
+- machinedeployments
+- machinesets
+- machines
+- machinehealthchecks
+- kubeadmcontrolplanes
+
+# metricAllowlist:
+#   - capi_.*
```
Comment (chrischdi) [inline hack/observability/capi-state-metrics/kube-state-metrics/values.yaml:51]: remove

File: hack/observability/capi-state-metrics/kube-state-metrics/values.yaml [hunk 3]
```diff
@@ -0,0 +1,65 @@
+# Default values for kube-state-metrics.
+prometheusScrape: true
+image:
+  # repository: registry.k8s.io/kube-state-metrics/kube-state-metrics
+  # tag: v2.5.0
+  # custom image which includes changes > v2.5.0 required for custom resource metrics
+  repository: chrischdi/kube-state-metrics
+  tag: v2.5.0-c766f5cb
```
Comment (chrischdi) [inline hack/observability/capi-state-metrics/kube-state-metrics/values.yaml:8]: TODO: update as soon as new release is out

File: Tiltfile (hunk 1 above)
Comment (killianmuldoon) [reply Tiltfile:359]: We don't need to expose this on localhost though, right? We can just make it available through a service in the cluster instead of needing a port-forward, so the port picked shouldn't really matter. Am I missing something?

Comment (chrischdi) [reply Tiltfile:359]: Good point :-) It is already available inside the cluster and gets scraped by prometheus 👍

File: hack/observability/capi-state-metrics/kustomization.yaml [hunk 4]
```diff
@@ -0,0 +1,53 @@
+resources:
```
Comment (sbueringer) [inline hack/observability/capi-state-metrics/kustomization.yaml:1]: Why do we have to rename?

**Review by sbueringer — COMMENTED** [review]
Just a few initial questions

File: hack/observability/capi-state-metrics/crd-config.yaml [hunk 5]
```diff
@@ -0,0 +1,656 @@
+kind: CustomResourceStateMetrics
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply hack/observability/capi-state-metrics/crd-config.yaml:1]: Yeah, still working on it. At least to get https://github.com/kubernetes/kube-state-metrics/pull/1810 in which would allow us to split the file on a per CR base.

File: hack/observability/capi-state-metrics/kustomization.yaml (hunk 4 above)
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kustomization.yaml:1]: We don't have to, I for now just did because it is limitted to the CAPI CR metrics for now. the other `kube_` metrics won't get exposed for now (and thus save memory etc.).

File: hack/observability/capi-state-metrics/kube-state-metrics/values.yaml (hunk 3 above)
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kube-state-metrics/values.yaml:8]: exactly

File: hack/observability/capi-state-metrics/kustomization.yaml (hunk 4 above)
Comment (sbueringer) [reply hack/observability/capi-state-metrics/kustomization.yaml:1]: But I assume we can eventually just use one kube-state-metrics directly from the chart with kube_ + our metrics?

File: hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+helmCharts:
+  - name: kube-state-metrics
+    repo: https://prometheus-community.github.io/helm-charts
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml:3]: Good point.

The prometheus chart re-uses this helm-chart and bundles it.

So if we want, we could integrate all this to the prometheus chart (and into `deploy_prometheus`) instead if wanted, or define it separately.

File: hack/observability/capi-state-metrics/kustomization.yaml (hunk 4 above)
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kustomization.yaml:1]: Yes it is possible.

File: hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml:3]: From user perspective: somebody may want to only provision kube-state-metrics without prometheus 🤔

[the remark this answers is not in the document]
Comment (sbueringer) [reply hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml:3]: And then retrieve and use the metrics manually?  :)

I wouldn't worry about this use case. It's also a dev env where we don't have to support everything.

But if you prefer to keep it separate that's fine for me

File: hack/observability/capi-state-metrics/kustomization.yaml (hunk 4 above)
Comment (sbueringer) [reply hack/observability/capi-state-metrics/kustomization.yaml:1]: I would prefer one kube-state-metrics with all metrics as the end-state of this PR (I realize it's WIP and you're using it to upstream stuff :))

File: hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml:3]: I was thinking from another perspective. Someone may have already deployed Capi and his own Prometheus setup.

And as addition now may want to deploy the Capi flavoured kube-state-metrics (and would be able to use `kustomize template` to this dir to generate the yams).

File: hack/observability/capi-state-metrics/kustomization.yaml (hunk 4 above)
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kustomization.yaml:1]: 👍

File: hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml:3]: Not sure if we have to cover use cases like this with our test-env deployments (and "support" them going forward). For me personally it's good enough if folks can build there own based on our example.

File: hack/observability/capi-state-metrics/crd-config.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrischdi) [reply hack/observability/capi-state-metrics/crd-config.yaml]: sadly this pr did not make it as I thought. Still only a single file for now. Future kube-state-metrics (ksm) may support config from a ksm-specific CR instead of a static config file. But that needs to be implemented.

Fallback to single file for now.
I added a ugly shell script to generate it (`metrics/build.sh`), ok for now because this PR is WIP. Todo: find a solution for that. Kustomize is not able to help here :-(

File: hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kube-state-metrics/kustomization.yaml]: I'd keep it as seperate for now. (do one thing but do it well).

File: hack/observability/kube-state-metrics/chart/values.yaml [hunk 7]
```diff
@@ -0,0 +1,82 @@
+image:
+  # TODO(chrischdi): drop to default to released image as soon as it got released
+  # repository: registry.k8s.io/kube-state-metrics/kube-state-metrics
+  # tag: v2.5.0
+  # custom image which includes changes > v2.5.0 required for custom resource metrics
+  repository: chrischdi/kube-state-metrics
+  tag: v2.5.0-fe097b6c
+  pullPolicy: IfNotPresent
+
```
Comment (chrischdi) [inline hack/observability/kube-state-metrics/chart/values.yaml:9]: TODO: remove as soon as ksm release + helm chart release is out

```suggestion
```

File: hack/observability/capi-state-metrics/kube-state-metrics/values.yaml (hunk 3 above)
Comment (chrischdi) [reply hack/observability/capi-state-metrics/kube-state-metrics/values.yaml]: moved over to https://github.com/kubernetes-sigs/cluster-api/pull/7095/files#r953778582

File: hack/observability/capi-state-metrics/crd-config.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (sbueringer) [reply hack/observability/capi-state-metrics/crd-config.yaml]: I think it's fine for the dev-env. Even splitting it in multiple files won't fix the big problem on how to keep that big of a YAML sync with our CRDs.

I think the eventual solution to this problem has to be a gen tool (but of course just a follow-up and in the "fullness of time" :))

[the remark this answers is not in the document]
Comment (chrischdi) [reply hack/observability/capi-state-metrics/crd-config.yaml]: ack 👍

**Comment by chrischdi** [comment]
Updated and I think its ready for review.

The next kube-state-metrics release should happen by tomorrow, so the last open point could get adressed afterwards.

Happy to receive feedback 👍 

Some first examples what could be queried:
* Get all resources which are paused (either by annotation or spec):

    <details>

    ```promql
    count(label_replace((
      capi_cluster_annotation_paused == 1 or capi_cluster_spec_paused == 1
      or capi_kubeadmcontrolplane_annotation_paused == 1
      or capi_machinedeployment_annotation_paused == 1 or capi_machinedeployment_spec_paused == 1
      or capi_machineset_annotation_paused == 1
      or capi_machine_annotation_paused == 1
      or capi_machinehealthcheck_annotation_paused == 1
    ), "object", "$1", "__name__", "capi_([a-z]+)_.*")) by (name,namespace,cluster_name,object)
    ```

    </details>

* First example Grafana dashboard (**Disclamer**: far from perfect! WIP but the current state):
    image


    <details>

    ```json
    {
      "__inputs": [
        {
          "name": "DS_PROMETHEUS",
          "label": "Prometheus",
          "description": "",
          "type": "datasource",
          "pluginId": "prometheus",
          "pluginName": "Prometheus"
        }
      ],
      "__elements": [],
      "__requires": [
        {
          "type": "grafana",
          "id": "grafana",
          "name": "Grafana",
          "version": "8.4.5"
        },
        {
          "type": "panel",
          "id": "piechart",
          "name": "Pie chart",
          "version": ""
        },
        {
          "type": "datasource",
          "id": "prometheus",
          "name": "Prometheus",
          "version": "1.0.0"
        },
        {
          "type": "panel",
          "id": "table",
          "name": "Table",
          "version": ""
        }
      ],
      "annotations": {
        "list": [
          {
            "builtIn": 1,
            "datasource": "-- Grafana --",
            "enable": true,
            "hide": true,
            "iconColor": "rgba(0, 211, 255, 1)",
            "name": "Annotations & Alerts",
            "target": {
              "limit": 100,
              "matchAny": false,
              "tags": [],
              "type": "dashboard"
            },
            "type": "dashboard"
          }
        ]
      },
      "editable": true,
      "fiscalYearStartMonth": 0,
      "graphTooltip": 0,
      "id": null,
      "iteration": 1661416446097,
      "links": [],
      "liveNow": false,
      "panels": [
        {
          "datasource": {
            "type": "prometheus",
            "uid": "${DS_PROMETHEUS}"
          },
          "description": "Lists of all Clusters",
          "fieldConfig": {
            "defaults": {
              "color": {
                "mode": "thresholds"
              },
              "custom": {
                "align": "auto",
                "displayMode": "auto"
              },
              "mappings": [],
              "thresholds": {
                "mode": "absolute",
                "steps": [
                  {
                    "color": "green",
                    "value": null
                  },
                  {
                    "color": "red",
                    "value": 80
                  }
                ]
              }
            },
            "overrides": [
              {
                "matcher": {
                  "id": "byName",
                  "options": "cluster_name"
                },
                "properties": [
                  {
                    "id": "custom.width",
                    "value": 341
                  }
                ]
              }
            ]
          },
          "gridPos": {
            "h": 8,
            "w": 14,
            "x": 0,
            "y": 0
          },
          "id": 3,
          "options": {
            "footer": {
              "fields": "",
              "reducer": [
                "sum"
              ],
              "show": false
            },
            "showHeader": true,
            "sortBy": []
          },
          "pluginVersion": "8.4.5",
          "targets": [
            {
              "datasource": {
                "type": "prometheus",
                "uid": "${DS_PROMETHEUS}"
              },
              "exemplar": false,
              "expr": "count(capi_cluster_info{namespace=~\"$namespace\",name=~\"$cluster\"}) by (namespace,name,uid,topology_class,topology_version)",
              "format": "table",
              "instant": true,
              "interval": "",
              "legendFormat": "",
              "refId": "A"
            }
          ],
          "title": "Clusters",
          "transformations": [
            {
              "id": "seriesToColumns",
              "options": {
                "byField": "uid"
              }
            },
            {
              "id": "organize",
              "options": {
                "excludeByName": {
                  "Time": true,
                  "Value": true
                },
                "indexByName": {
                  "Time": 0,
                  "Value": 6,
                  "cluster_name": 4,
                  "name": 3,
                  "namespace": 2,
                  "resource": 1,
                  "uid": 5
                },
                "renameByName": {}
              }
            }
          ],
          "type": "table"
        },
        {
          "datasource": {
            "type": "prometheus",
            "uid": "${DS_PROMETHEUS}"
          },
          "description": "",
          "fieldConfig": {
            "defaults": {
              "color": {
                "mode": "palette-classic"
              },
              "custom": {
                "hideFrom": {
                  "legend": false,
                  "tooltip": false,
                  "viz": false
                }
              },
              "mappings": []
            },
            "overrides": []
          },
          "gridPos": {
            "h": 8,
            "w": 5,
            "x": 14,
            "y": 0
          },
          "id": 5,
          "options": {
            "displayLabels": [
              "name"
            ],
            "legend": {
              "displayMode": "list",
              "placement": "right",
              "values": []
            },
            "pieType": "pie",
            "reduceOptions": {
              "calcs": [
                "lastNotNull"
              ],
              "fields": "",
              "values": false
            },
            "tooltip": {
              "mode": "single",
              "sort": "none"
            }
          },
          "pluginVersion": "8.4.5",
          "targets": [
            {
              "datasource": {
                "type": "prometheus",
                "uid": "${DS_PROMETHEUS}"
              },
              "exemplar": false,
              "expr": "count(capi_cluster_info{namespace=~\"$namespace\",name=~\"$cluster\"}) by (topology_class)",
              "format": "time_series",
              "instant": true,
              "interval": "",
              "legendFormat": "{{ topology_class }}",
              "refId": "A"
            }
          ],
          "title": "Clusters by Topology Classes",
          "transformations": [],
          "type": "piechart"
        },
        {
          "datasource": {
            "type": "prometheus",
            "uid": "${DS_PROMETHEUS}"
          },
          "description": "",
          "fieldConfig": {
            "defaults": {
              "color": {
                "fixedColor": "blue",
                "mode": "palette-classic"
              },
              "custom": {
                "hideFrom": {
                  "legend": false,
                  "tooltip": false,
                  "viz": false
                }
              },
              "mappings": []
            },
            "overrides": []
          },
          "gridPos": {
            "h": 8,
            "w": 5,
            "x": 19,
            "y": 0
          },
          "id": 4,
          "options": {
            "displayLabels": [
              "name"
            ],
            "legend": {
              "displayMode": "list",
              "placement": "right",
              "values": []
            },
            "pieType": "pie",
            "reduceOptions": {
              "calcs": [
                "lastNotNull"
              ],
              "fields": "",
              "values": false
            },
            "tooltip": {
              "mode": "single",
              "sort": "none"
            }
          },
          "pluginVersion": "8.4.5",
          "targets": [
            {
              "datasource": {
                "type": "prometheus",
                "uid": "${DS_PROMETHEUS}"
              },
              "exemplar": false,
              "expr": "count(capi_cluster_info{namespace=~\"$namespace\",name=~\"$cluster\"}) by (topology_version)",
              "format": "time_series",
              "instant": true,
              "interval": "",
              "legendFormat": "{{ topology_version }}",
              "refId": "A"
            }
          ],
          "title": "Clusters by Topology Version",
          "transformations": [],
          "type": "piechart"
        },
        {
          "collapsed": false,
          "gridPos": {
            "h": 1,
            "w": 24,
            "x": 0,
            "y": 8
          },
          "id": 9,
          "panels": [],
          "title": "Paused resources",
          "type": "row"
        },
        {
          "datasource": {
            "type": "prometheus",
            "uid": "${DS_PROMETHEUS}"
          },
          "description": "Lists all paused CAPI resources",
          "fieldConfig": {
            "defaults": {
              "color": {
                "mode": "thresholds"
              },
              "custom": {
                "align": "auto",
                "displayMode": "auto"
              },
              "mappings": [],
              "thresholds": {
                "mode": "absolute",
                "steps": [
                  {
                    "color": "green",
                    "value": null
                  },
                  {
                    "color": "red",
                    "value": 80
                  }
                ]
              }
            },
            "overrides": [
              {
                "matcher": {
                  "id": "byName",
                  "options": "cluster_name"
                },
                "properties": [
                  {
                    "id": "custom.width",
                    "value": 341
                  }
                ]
              }
            ]
          },
          "gridPos": {
            "h": 11,
            "w": 24,
            "x": 0,
            "y": 9
          },
          "id": 2,
          "options": {
            "footer": {
              "fields": "",
              "reducer": [
                "sum"
              ],
              "show": false
            },
            "showHeader": true,
            "sortBy": []
          },
          "pluginVersion": "8.4.5",
          "targets": [
            {
              "datasource": {
                "type": "prometheus",
                "uid": "${DS_PROMETHEUS}"
              },
              "exemplar": false,
              "expr": "count(label_replace((\n  capi_cluster_annotation_paused{namespace=~\"$namespace\",name=~\"$cluster\"} == 1 or capi_cluster_spec_paused{namespace=~\"$namespace\",name=~\"$cluster\"} == 1\n  or capi_kubeadmcontrolplane_annotation_paused{namespace=~\"$namespace\",cluster_name=~\"$cluster\"} == 1\n  or capi_machinedeployment_annotation_paused{namespace=~\"$namespace\",cluster_name=~\"$cluster\"} == 1 or capi_machinedeployment_spec_paused{namespace=~\"$namespace\",cluster_name=~\"$cluster\"} == 1\n  or capi_machineset_annotation_paused{namespace=~\"$namespace\",cluster_name=~\"$cluster\"} == 1\n  or capi_machine_annotation_paused{namespace=~\"$namespace\",cluster_name=~\"$cluster\"} == 1\n  or capi_machinehealthcheck_annotation_paused{namespace=~\"$namespace\",cluster_name=~\"$cluster\"} == 1\n), \"resource\", \"$1\", \"__name__\", \"capi_([a-z]+)_.*\")) by (name,namespace,cluster_name,resource,uid)",
              "format": "table",
              "instant": true,
              "interval": "",
              "legendFormat": "",
              "refId": "A"
            }
          ],
          "title": "Paused resources",
          "transformations": [
            {
              "id": "seriesToColumns",
              "options": {
                "byField": "uid"
              }
            },
            {
              "id": "organize",
              "options": {
                "excludeByName": {
                  "Time": true,
                  "Value": true
                },
                "indexByName": {
                  "Time": 0,
                  "Value": 6,
                  "cluster_name": 4,
                  "name": 3,
                  "namespace": 2,
                  "resource": 1,
                  "uid": 5
                },
                "renameByName": {}
              }
            }
          ],
          "type": "table"
        }
      ],
      "refresh": "",
      "schemaVersion": 35,
      "style": "dark",
      "tags": [],
      "templating": {
        "list": [
          {
            "allValue": ".*",
            "current": {},
            "datasource": {
              "type": "prometheus",
              "uid": "${DS_PROMETHEUS}"
            },
            "definition": "label_values(capi_cluster_info, namespace)",
            "hide": 0,
            "includeAll": true,
            "label": "Namespace",
            "multi": true,
            "name": "namespace",
            "options": [],
            "query": {
              "query": "label_values(capi_cluster_info, namespace)",
              "refId": "StandardVariableQuery"
            },
            "refresh": 1,
            "regex": "",
            "skipUrlSync": false,
            "sort": 0,
            "type": "query"
          },
          {
            "allValue": ".*",
            "current": {},
            "datasource": {
              "type": "prometheus",
              "uid": "${DS_PROMETHEUS}"
            },
            "definition": "label_values(capi_cluster_info{namespace=~\"$namespace\"}, name)",
            "hide": 0,
            "includeAll": true,
            "label": "Cluster",
            "multi": true,
            "name": "cluster",
            "options": [],
            "query": {
              "query": "label_values(capi_cluster_info{namespace=~\"$namespace\"}, name)",
              "refId": "StandardVariableQuery"
            },
            "refresh": 1,
            "regex": "",
            "skipUrlSync": false,
            "sort": 0,
            "type": "query"
          }
        ]
      },
      "time": {
        "from": "now-6h",
        "to": "now"
      },
      "timepicker": {},
      "timezone": "",
      "title": "Cluster API Overview (WIP)",
      "uid": "qiW7XdZVkasd",
      "version": 5,
      "weekStart": ""
    }
    ```
    
    </details>
   
/retitle ✨ Integrate kube-state-metrics and CR config into tilt.

File: hack/observability/kube-state-metrics/crd-config.yaml [hunk 8]
```diff
@@ -0,0 +1,745 @@
...
+      namespace:
+      - metadata
+      - namespace
+      uid:
+      - metadata
+      - uid
+    metricNamePrefix: capi_cluster
+    metrics:
+    - name: info
+      help: Information.
```
Comment (fabriziopandini) [inline hack/observability/kube-state-metrics/crd-config.yaml:21]: When this PR merge let's open a follow up issue to improve metrics descriptions (most of them are basic or TODO)

File: hack/observability/kube-state-metrics/crd-config.yaml [hunk 9]
```diff
@@ -0,0 +1,745 @@
+kind: CustomResourceStateMetrics
```
Comment (fabriziopandini) [inline hack/observability/kube-state-metrics/crd-config.yaml:1]: Is it possible to add an header comment specifying this is an autogenerated file?

File: hack/observability/kube-state-metrics/metrics/build.sh [hunk 10]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
```
Comment (fabriziopandini) [inline hack/observability/kube-state-metrics/metrics/build.sh:1]: Let's add a readme file in this directory with instructions about how to generate the metrics file.
It could also include a disclaimer specifying this is temporary and the long-term goal is to have method to auto-generate metric configuration from API type markers

**Review by fabriziopandini — COMMENTED** [review]
Great to see this nearing the finish line!
Overall lgtm, I still need to make a detaied pass to all the metrics

File: hack/observability/kube-state-metrics/crd-config.yaml (hunk 8 above)
Comment (chrischdi) [reply hack/observability/kube-state-metrics/crd-config.yaml:21]: ah, missed that and wanted to add help everywhere :-( will update it :-)

File: docs/book/src/developer/tilt.md [hunk 11]
```diff
@@ -173,9 +173,20 @@ kustomize_substitutions:
...
+Supported values are:
+
+  * `grafana`*: To create dashboards and query `loki` as well as `prometheus`.
+  * `kube-state-metrics`: For exposing metrics for kubernetes and CAPI resources to `prometheus`.
+  * `loki`: To receive and store logs.
+  * `prometheus`: For collecting metrics from Kubernetes..
+  * `promtail`: For providing pod logs to `loki`.
+  * `visualizer`: Visualize all the Cluster API custom resources for each cluster, provide quick access to the specs and status of any resource.
+
+\*: Note: the UI will be accessible via a link in the tilt console
```
[the remark this answers is not in the document]
Comment (chrischdi) [reply docs/book/src/developer/tilt.md:188]: ah no, I missed saving the file in the IDE and forgot to add the star for `prometheus`, `grafana` and `visualizer`. It applies to all of them and is referenced via the star. Should be fixed now.

File: hack/observability/kube-state-metrics/crd-config.yaml (hunk 8 above)
Comment (chrischdi) [reply hack/observability/kube-state-metrics/crd-config.yaml]: Updated all metrics to contain help information. Texts got adapted from existing kube-state-metrics help information.

File: hack/observability/kube-state-metrics/crd-config.yaml (hunk 9 above)
Comment (chrischdi) [reply hack/observability/kube-state-metrics/crd-config.yaml:3]: Added a comment via the header file :-)

File: docs/book/src/developer/tilt.md [hunk 12]
```diff
@@ -173,9 +173,20 @@ kustomize_substitutions:
...
 Important! This feature requires the `helm` command to be available in the user's path.
 
+Supported values are:
+
+  * `grafana`*: To create dashboards and query `loki` as well as `prometheus`.
+  * `kube-state-metrics`: For exposing metrics for kubernetes and CAPI resources to `prometheus`.
+  * `loki`: To receive and store logs.
+  * `prometheus`*: For collecting metrics from Kubernetes..
+  * `promtail`: For providing pod logs to `loki`.
+  * `visualizer`*: Visualize all the Cluster API custom resources for each cluster, provide quick access to the specs and status of any resource.
```
Comment (sbueringer) [inline docs/book/src/developer/tilt.md:186]:
```suggestion
  * `visualizer`*: Visualize Cluster API resources for each cluster, provide quick access to the specs and status of any resource.
```
nit: I just think it sounds better that way, up to you

File: docs/proposals/20220411-cluster-api-state-metrics.md [hunk 13]
```diff
@@ -10,7 +10,7 @@ reviewers:
   - "@apricote"
   - "@fabriziopandini"
 creation-date: 2022-03-03
-last-updated: 2022-03-15
+last-updated: 2022-08-25
 status: experimental
 ---
```
Comment (sbueringer) [inline docs/proposals/20220411-cluster-api-state-metrics.md:15]: Should we maybe do this in a separate PR? I think there is more outdated in this proposal and I don't really want to block this PR because of that

File: hack/observability/kube-state-metrics/metrics/regen.sh [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Would be great to have a generator replacing this script and yaml files.
+
+set -e
+
+DIR="$(dirname ${BASH_SOURCE[0]})"
```
Comment (sbueringer) [inline hack/observability/kube-state-metrics/metrics/regen.sh:21]: Should we call this script in some gen + verify-gen target?

File: hack/observability/kube-state-metrics/kustomization.yaml [hunk 15]
```diff
@@ -0,0 +1,12 @@
+resources:
+  - ../namespace.yaml
+  # The kube-state-metrics chart needs to be referenced in a sub-kustomization.yaml,
```
Comment (sbueringer) [inline hack/observability/kube-state-metrics/kustomization.yaml:3]: How does this work exactly? Are we overwriting a ConfigMap from the chart?

Comment (chrischdi) [reply hack/observability/kube-state-metrics/kustomization.yaml:3]: Nope.

The ConfigMapGenerator below creates a configmap with name `kube-state-metrics-crd-config` which results in a configmap called `kube-state-metrics-crd-config-<suffix>`.

The helm chart has references to a configmap named `kube-state-metrics-crd-config` (which is the same as our configmapgenerator) (it does not have the configmap, just the volume mount reference inside the deployment).

Kustomize then replaces the referenced name in the deployments volume reference from `kube-state-metrics-crd-config` to `kube-state-metrics-crd-config-<suffix>`, where suffix is deterministic from the content of the configmap.

This functionality would for every yaml referenced by an entry in `resources`, but does not work if we would move the content of `hack/observability/kube-state-metrics/chart/kustomization.yaml` to here.

File: docs/proposals/20220411-cluster-api-state-metrics.md (hunk 13 above)
Comment (chrischdi) [reply docs/proposals/20220411-cluster-api-state-metrics.md:15]: I'd be ok with that, but unsure when merging this then first and updating the proposal afterwards 🤔?!

File: hack/observability/kube-state-metrics/metrics/regen.sh (hunk 14 above)
Comment (chrischdi) [reply hack/observability/kube-state-metrics/metrics/regen.sh:21]: Sounds better 👍 

I moved the script to a make target and removed it and added it to the `generate` target. This way it is also included in `verify-generate`.

**Comment by chrischdi** [comment]
/hold

pending squash + reviews :-)

File: docs/proposals/20220411-cluster-api-state-metrics.md (hunk 13 above)
Comment (sbueringer) [reply docs/proposals/20220411-cluster-api-state-metrics.md:15]: I think it's fine to update the proposal in parallel/separate/afterwards. It wouldn't be the first time we update proposals after we changed the implementation.

We also had clear consensus on the issue that we want to implement it directly in kube-state-metrics instead. So I think we don't need the proposal update PR as some kind of approval.

File: hack/observability/kube-state-metrics/metrics/regen.sh (hunk 14 above)
Comment (sbueringer) [reply hack/observability/kube-state-metrics/metrics/regen.sh]: Thx!

File: hack/observability/kube-state-metrics/metrics/header.yaml [hunk 16]
```diff
@@ -0,0 +1,5 @@
+# This file was auto-generated via ./hack/observability/kube-state-metrics/metrics/regen.sh.
```
Comment (sbueringer) [inline hack/observability/kube-state-metrics/metrics/header.yaml:1]:
```suggestion
# This file was auto-generated via make generate-metrics-config
```
nit

File: hack/observability/kube-state-metrics/kustomization.yaml (hunk 15 above)
Comment (sbueringer) [reply hack/observability/kube-state-metrics/kustomization.yaml:3]: Ah perfect thx.

Please consider expanding the current comment. I think without background knowledge it's hard to infer that based on the current comment.

File: docs/proposals/20220411-cluster-api-state-metrics.md (hunk 13 above)
Comment (chrischdi) [reply docs/proposals/20220411-cluster-api-state-metrics.md:15]: Removed the changes for this file from the PR. I will create a separate one for it 👍

**Comment by fabriziopandini** [comment]
Made a pass on the metrics, and I think this is a good set to start with. 
Looking forward to start using those in dashboards so we can better appreciate what is useful and if something else is missing

File: hack/observability/kube-state-metrics/metrics/header.yaml [hunk 17]
```diff
@@ -0,0 +1,5 @@
+# This file was auto-generated via: make generate-metrics-config
```
Comment (sbueringer) [inline hack/observability/kube-state-metrics/metrics/header.yaml:1]: This file seems not to be generated?

File: hack/observability/kube-state-metrics/kustomization.yaml (hunk 15 above)
Comment (sbueringer) [reply hack/observability/kube-state-metrics/kustomization.yaml]: Perfect, thx!

**Comment by sbueringer** [comment]
lgtm pending https://github.com/kubernetes-sigs/cluster-api/pull/7095#discussion_r953778582

File: hack/observability/kube-state-metrics/chart/values.yaml [hunk 18]
```diff
@@ -0,0 +1,77 @@
+image:
+  tag: v2.6.0
```
Comment (sbueringer) [inline hack/observability/kube-state-metrics/chart/values.yaml:2]: Would be good if we can drop the pinned tag once the corresponding Helm chart is out
(just that when we pull the latest chart the pinned tag and the chart don't run out of sync)

**Comment by sbueringer** [comment]
/lgtm

@chrischdi hold cancel?

File: hack/observability/kube-state-metrics/chart/values.yaml (hunk 18 above)
Comment (chrischdi) [reply hack/observability/kube-state-metrics/chart/values.yaml:2]: I agree, the latest chart still refers 2.5.0

**Comment by sbueringer** [comment]
/assign @fabriziopandini 
for final review

**Comment by killianmuldoon** [comment]
/lgtm
This is awesome, and working well on my machine. 
One question about future work - what's the scope for including other providers objects in this? Is there an easy way for use to point to provider metrics definitions e.g. CAPD and have those loaded? Or is the idea that end users would load in their own metrics files and re-generate the metrics?

File: hack/observability/kube-state-metrics/chart/values.yaml (hunk 18 above)
Comment (fabriziopandini) [reply hack/observability/kube-state-metrics/chart/values.yaml:2]: Let's open an issue to track this

**Comment by fabriziopandini** [comment]
/lgtm
/approve

> One question about future work - what's the scope for including other providers objects in this? Is there an easy way for use to point to provider metrics definitions e.g. CAPD and have those loaded? Or is the idea that end users would load in their own metrics files and re-generate the metrics?

I think there is an intermediate step to get there, that is to automate the generation of  the metrics configuration by introducing a new set of markers; this will make adoption in the providers straight forward

**Comment by sbueringer** [comment]
/hold cancel
Thx @chrischdi 
Really nice work, super super useful! :))

File: hack/observability/kube-state-metrics/chart/values.yaml (hunk 18 above)
Comment (sbueringer) [reply hack/observability/kube-state-metrics/chart/values.yaml:2]: done: #7143
